Port neural decision forest

anas-rz opened the pull request:
[Colab Notebook](https://colab.research.google.com/drive/1EvSmz35vZL1eK6dQ9I_hhem92qKnSOBO?usp=sharing)

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,459 @@
+    for feature_name in FEATURE_NAMES:
+        if feature_name in NUMERIC_FEATURE_NAMES:
+            inputs[feature_name] = layers.Input(
+                name=feature_name, shape=(), dtype=tf.float32
```
Prefer using `"float32"`

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,459 @@
+            )
+        else:
+            inputs[feature_name] = layers.Input(
+                name=feature_name, shape=(), dtype=tf.string
```
Prefer using `"string"`

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,459 @@
+        self.used_features_mask = one_hot[sampled_feature_indicies]
+
+        # Initialize the weights of the classes in leaves.
+        self.pi = tf.Variable(
```
I'm pretty sure this won't work -- the variable won't be tracked. Use `self.add_weight` instead.

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,459 @@
+## Setup
+"""
+
+import keras_core as keras
```
Set the backend environment variable first

fchollet reviewed (commented):
Thanks for the PR!

Porting this to backend-agnostic will require moving string preprocessing out of the model.

anas-rz commented:
Moved preprocessing away from model and converted all operations from tf to ops. 
It works with TensorFlow.
One problem I face again and again with PyTorch backend is data type conversion. It doesn't automatically infer the datatypes. The code doesn't work with jax as it doesn't identify add_weight as a variable `self.pi`. The COLAB notebook in the first comment can be used to reproduce the error.

fchollet commented:
> One problem I face again and again with PyTorch backend is data type conversion. It doesn't automatically infer the datatypes. The code doesn't work with jax as it doesn't identify add_weight as a variable self.pi. 

Can you expand? What's the issue?

anas-rz commented:
I have faced this type casting issue on many examples `expected scalar type Float but found Double` with PyTorch backend. While using pure PyTorch I solve this kind of error by calling .float() methods on tensors. 

For this particular example, the error thrown with JAX Backend is this for self.pi weights:

`Argument '<KerasVariable shape=(1024, 2), dtype=float32, name=variable>' of type <class 'keras_core.src.backend.jax.core.Variable'> is not a valid JAX type.`

fchollet commented:
> For this particular example, the error thrown with JAX Backend is this for self.pi weights:

Can you post the full stack trace?

anas-rz commented:
Apparently the error comes from applying softmax to the backend variable.

`TypeError                                 Traceback (most recent call last)
[<ipython-input-3-01c36cde8315>](https://localhost:8080/#) in <cell line: 423>()
    421 
    422 tree_model = create_tree_model()
--> 423 run_experiment(tree_model)
    424 
    425 

6 frames
[<ipython-input-3-01c36cde8315>](https://localhost:8080/#) in run_experiment(model)
    384     )
    385 
--> 386     model.fit(train_dataset, epochs=num_epochs)
    387     print("Model training finished")
    388 

[/usr/local/lib/python3.10/dist-packages/keras_core/src/utils/traceback_utils.py](https://localhost:8080/#) in error_handler(*args, **kwargs)
    121             # To get the full stack trace, call:
    122             # `keras_core.config.disable_traceback_filtering()`
--> 123             raise e.with_traceback(filtered_tb) from None
    124         finally:
    125             del filtered_tb

    [... skipping hidden 20 frame]

[<ipython-input-3-01c36cde8315>](https://localhost:8080/#) in call(self, features)
    326 
    327         mu = ops.reshape(mu, [batch_size, self.num_leaves])  # [batch_size, num_leaves]
--> 328         probabilities = keras.activations.softmax(self.pi)  # [num_leaves, num_classes]
    329         outputs = ops.matmul(mu, probabilities)  # [batch_size, num_classes]
    330         return outputs

[/usr/local/lib/python3.10/dist-packages/jax/_src/nn/functions.py](https://localhost:8080/#) in softmax(x, axis, where, initial)
    350     return _softmax(x, axis, where, initial)
    351   else:
--> 352     return _softmax_deprecated(x, axis, where, initial)
    353 
    354 # TODO(mattjj): replace softmax with _softmax when deprecation flag is removed

[/usr/local/lib/python3.10/dist-packages/jax/_src/nn/functions.py](https://localhost:8080/#) in _softmax_deprecated(x, axis, where, initial)
    373 
    374 def _softmax_deprecated(x, axis, where, initial):
--> 375   x_max = jnp.max(x, axis, where=where, initial=initial, keepdims=True)
    376   unnormalized = jnp.exp(x - lax.stop_gradient(x_max))
    377   result = unnormalized / jnp.sum(unnormalized, axis, where=where, keepdims=True)

[/usr/local/lib/python3.10/dist-packages/jax/_src/numpy/reductions.py](https://localhost:8080/#) in max(a, axis, out, keepdims, initial, where)
    254         keepdims: bool = False, initial: Optional[ArrayLike] = None,
    255         where: Optional[ArrayLike] = None) -> Array:
--> 256   return _reduce_max(a, axis=_ensure_optional_axes(axis), out=out,
    257                      keepdims=keepdims, initial=initial, where=where)
    258 

    [... skipping hidden 3 frame]

[/usr/local/lib/python3.10/dist-packages/jax/_src/dispatch.py](https://localhost:8080/#) in check_arg(arg)
    400 def check_arg(arg):
    401   if not (isinstance(arg, core.Tracer) or _valid_jaxtype(arg)):
--> 402     raise TypeError(f"Argument '{arg}' of type {type(arg)} is not a valid "
    403                     "JAX type.")
    404 

TypeError: Exception encountered when calling NeuralDecisionTree.call().

Argument '<KerasVariable shape=(1024, 2), dtype=float32, name=variable>' of type <class 'keras_core.src.backend.jax.core.Variable'> is not a valid JAX type.

Arguments received by NeuralDecisionTree.call():
  • features=jnp.ndarray(shape=(265, 53), dtype=float32)`

fchollet commented:
Thanks for posting the trace. This should now be fixed at HEAD -- please check.

anas-rz commented:
Tried pulling and installing the latest version from GitHub and it works with JAX backend. For the PyTorch backend, it throws the same error with full stack trace:
```
RuntimeError                              Traceback (most recent call last)
[<ipython-input-2-aa228ae06391>](https://localhost:8080/#) in <cell line: 423>()
    421 
    422 tree_model = create_tree_model()
--> 423 run_experiment(tree_model)
    424 
    425 

4 frames
[<ipython-input-2-aa228ae06391>](https://localhost:8080/#) in run_experiment(model)
    384     )
    385 
--> 386     model.fit(train_dataset, epochs=num_epochs)
    387     print("Model training finished")
    388 

[/usr/local/lib/python3.10/dist-packages/keras_core/src/utils/traceback_utils.py](https://localhost:8080/#) in error_handler(*args, **kwargs)
    121             # To get the full stack trace, call:
    122             # `keras_core.config.disable_traceback_filtering()`
--> 123             raise e.with_traceback(filtered_tb) from None
    124         finally:
    125             del filtered_tb

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *args, **kwargs)
   1499                 or _global_backward_pre_hooks or _global_backward_hooks
   1500                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501             return forward_call(*args, **kwargs)
   1502         # Do not call functions when jit is used
   1503         full_backward_hooks, non_full_backward_hooks = [], []

[/usr/local/lib/python3.10/dist-packages/torch/nn/modules/module.py](https://localhost:8080/#) in _call_impl(self, *args, **kwargs)
   1499                 or _global_backward_pre_hooks or _global_backward_hooks
   1500                 or _global_forward_hooks or _global_forward_pre_hooks):
-> 1501             return forward_call(*args, **kwargs)
   1502         # Do not call functions when jit is used
   1503         full_backward_hooks, non_full_backward_hooks = [], []

[<ipython-input-2-aa228ae06391>](https://localhost:8080/#) in call(self, features)
    298 
    299         # Apply the feature mask to the input features.
--> 300         features = ops.matmul(
    301             features, ops.transpose(self.used_features_mask)
    302         )  # [batch_size, num_used_features]

RuntimeError: Exception encountered when calling NeuralDecisionTree.call().

expected scalar type Float but found Double
```

fchollet commented:
I think we're just a `cast` away (in torch `matmul`) from making it work. Can you try printing the dtypes of `features` and `self.used_features_mask`?

anas-rz commented:
It says `float64`

anas-rz commented:
If I go to the backend/torch/numpy ops
The function casts both of them to tensor
```
def matmul(x1, x2):
    x1, x2 = convert_to_tensor(x1), convert_to_tensor(x2)
    return torch.matmul(x1, x2)
```

anas-rz commented:
ops.transpose converts too
```
def transpose(x, axes=None):
    x = convert_to_tensor(x)
    if axes is not None:
        return torch.permute(x, dims=axes)
    return x.T
```

fchollet commented:
We standardize on the NumPy API, and the NumPy will return float64 upon matmul with a float32 and float64 input. So I think we should also do this in our `backend.torch.numpy.matmul` op. One thing that intrigues me however is why the inputs are float32 and float64 in the first place. In TF this would have errored out. Are they different float types will all backends?

anas-rz commented:
It's the same for all backends. 
Jax Backend:
```
self.used_features_mask: float64
features: float32
```
TensorFlow Backend
```
self.used_features_mask: float64
features: <dtype: 'float32'>
```

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,471 @@
+        # Create a mask for the randomly selected features.
+        num_used_features = int(num_features * used_features_rate)
+        one_hot = np.eye(num_features)
+        sampled_feature_indicies = np.random.choice(
```
Typo: indices

fchollet commented on examples/keras_io/tensorflow/structured_data/deep_neural_decision_forests.py:
```diff
@@ -0,0 +1,471 @@
+        sampled_feature_indicies = np.random.choice(
+            np.arange(num_features), num_used_features, replace=False
+        )
+        self.used_features_mask = one_hot[sampled_feature_indicies]
```
This is float64 because it's a numpy array. You could just do `convert_to_tensor` on it to fix the issue.

fchollet commented:
> The types printed for different backends are in different format.

I think that's just because one is a variable and the other is a tensor. It's normal. Thanks for posting the dtypes.

fchollet reviewed (commented):
LGTM -- does it work with all backends now?

anas-rz commented:
Yes.

anas-rz commented:
Yes

fchollet reviewed (approved):
LGTM -- thank you!
